include/glow/Base/TaggedList.h

```diff
@@ -0,0 +1,398 @@
+    node->nextTaggedNode_ = nullptr;
+    node->prevTaggedNode_ = nullptr;
+    prev->nextTaggedNode_ = next;
+    next->prevTaggedNode_ = prev;
+  }
+
+  void clear() {
+    sentinel_.prevTaggedNode_ = sentinel_.nextTaggedNode_ = &sentinel_;
+    size_ = 0;
+  }
+
+private:
+  void renumber(NodeBase *lo, NodeBase *hi);
+  NodeBase sentinel_;
+  // Not a size_t because the list can't hold more than 2^32 nodes.
+  std::uint32_t size_ = 0;
+};
+
+} // namespace tagged_list_details
+
+/// Base class for intrusive list nodes of type T.
+///
+/// A linked list node of type T must be derived from TaggedListNode<T>.
+template <typename T>
+class TaggedListNode : public tagged_list_details::NodeBase {
+public:
+  // Get an iterator pointing at this node.
+  auto getIterator() {
+    return tagged_list_details::Iterator<T, false, false>(this);
+  }
+
+  // Get a const iterator pointing at this node.
+  auto getIterator() const {
+    return tagged_list_details::Iterator<T, false, true>(this);
+  }
+
+  // Get a reverse iterator pointing at this node.
+  auto getReverseIterator() {
+    return tagged_list_details::Iterator<T, true, false>(this);
+  }
+
+  // Get a const reverse iterator pointing at this node.
+  auto getReverseIterator() const {
+    return tagged_list_details::Iterator<T, true, true>(this);
+  }
+};
+
+template <typename T, typename Traits = TaggedListTraits<T>>
+class TaggedList : public Traits {
+public:
+  using value_type = T;
+  using iterator = tagged_list_details::Iterator<T, false, false>;
+  using const_iterator = tagged_list_details::Iterator<T, false, true>;
+  using reverse_iterator = tagged_list_details::Iterator<T, true, false>;
+  using const_reverse_iterator = tagged_list_details::Iterator<T, true, true>;
+
+  ~TaggedList() { clear(); }
+
+  /// Get the number of elements in the list.
+  /// This is a constant time operation.
+  size_t size() const { return impl_.size(); }
+
+  /// Is the list empty?
+  bool empty() const { return size() == 0; }
+
+  iterator begin() { return impl_.begin(); }
+  iterator end() { return impl_.end(); }
+  reverse_iterator rbegin() { return impl_.rbegin(); }
+  reverse_iterator rend() { return impl_.rend(); }
+  const_iterator begin() const { return impl_.begin(); }
+  const_iterator end() const { return impl_.end(); }
+  const_reverse_iterator rbegin() const { return impl_.begin(); }
+  const_reverse_iterator rend() const { return impl_.end(); }
+
+  /// Insert `node` before `next`.
+  ///
+  /// The node is not copied, and it can't already be on another list.
+  ///
+  /// Return s an iterator pointing to the inserted node.
+  iterator insert(iterator next, T *node) {
+    assert(node && !node->inTaggedList());
+    this->addNodeToList(node);
+    impl_.insert(next.node_, node);
+    return iterator(node);
+  }
+
+  /// Remove `node` from this list without deleting it.
+  ///
+  /// Returns a pointer to the removed node.
+  T *remove(T *node) {
+    assert(node && node->inTaggedList());
```

Comment:
Should it also check that the node belongs to the current list?

Reply:
That would be nice, but the way the abstraction works, a list node doesn't know which list it belongs to.

We have parent pointers in the `Instruction` nodes, but they are not available to the list implementation itself. See the hacks around `getContainingFunction()`.